### DISCUSSION TITLE
.NET 8 Preview. XamlGeneration source error

### DISCUSSION THREAD

**Discussion by limeniye** [unoplatform/uno#12848] [General]
Good day, Uno Platform team.
I notice that Uno Platform .Net 8 is not build without ```<UnoUIUseRoslynSourceGenerators>true</UnoUIUseRoslynSourceGenerators>``` parameter.

If I will change the UnoUIUseRoslynSourceGenerators parameter to false, I will catch following compilation error:

```
1>MSBUILD : error : Generation failed: System.AggregateException: One or more errors occurred. ---> System.InvalidOperationException: Generation failed for Uno.UI.SourceGenerators.XamlGenerator.XamlCodeGenerator. System.AggregateException: One or more errors occurred. ---> System.Exception: Processing failed for file C:\Users\mykyt\OneDrive\Робочий стіл\UnoAppNetEight\UnoAppNetEight\Presentation\Shell.xaml (System.ArgumentException: An item with the same key has already been added.
1>MSBUILD : error :    at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
1>MSBUILD : error :    at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
1>MSBUILD : error :    at Uno.UI.SourceGenerators.XamlGenerator.GenerationRunFileInfo.SetAppliedTypes(Dictionary`2 appliedTypes) in /home/vsts/work/1/s/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/GenerationRunFileInfo.cs:line 19
1>MSBUILD : error :    at Uno.UI.SourceGenerators.XamlGenerator.XamlFileGenerator.InnerGenerateFile() in /home/vsts/work/1/s/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:line 354
1>MSBUILD : error :    at Uno.UI.SourceGenerators.XamlGenerator.XamlFileGenerator.GenerateFile() in /home/vsts/work/1/s/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:line 315) ---> System.ArgumentException: An item with the same key has already been added.
1>MSBUILD : error :    at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
1>MSBUILD : error :    at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
1>MSBUILD : error :    at Uno.UI.SourceGenerators.XamlGenerator.GenerationRunFileInfo.SetAppliedTypes(Dictionary`2 appliedTypes) in /home/vsts/work/1/s/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/GenerationRunFileInfo.cs:line 19
1>MSBUILD : error :    at Uno.UI.SourceGenerators.XamlGenerator.XamlFileGenerator.InnerGenerateFile() in /home/vsts/work/1/s/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:line 490
1>MSBUILD : error :    at Uno.UI.SourceGenerators.XamlGenerator.XamlFileGenerator.GenerateFile() in /home/vsts/work/1/s/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:line 315
1>MSBUILD : error :    --- End of inner exception stack trace ---
1>MSBUILD : error :    at Uno.UI.SourceGenerators.XamlGenerator.XamlFileGenerator.GenerateFile() in /home/vsts/work/1/s/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:line 323
1>MSBUILD : error :    at Uno.UI.SourceGenerators.XamlGenerator.XamlCodeGeneration.<>c__DisplayClass178_0.<Generate>b__2(XamlFileDefinition file) in /home/vsts/work/1/s/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:line 444
1>MSBUILD : error :    at System.Linq.Parallel.PartitionedDataSource`1.ListContiguousIndexRangeEnumerator.MoveNext(T& currentElement, Int32& currentKey)
1>MSBUILD : error :    at System.Linq.Parallel.StopAndGoSpoolingTask`2.SpoolingWork()
1>MSBUILD : error :    at System.Linq.Parallel.SpoolingTaskBase.Work()
1>MSBUILD : error :    at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
1>MSBUILD : error :    at System.Threading.Tasks.Task.Execute()
```

I have a proposition to add this parameter as default with Uno Platform .Net 8 creating project.

Visual Studio: 17.7.0 Preview 2.0
Uno Platform: 4.9.17.18
Dotnet sdk-list:
```
7.0.400-preview.23274.1 [C:\Program Files\dotnet\sdk]
8.0.100-preview.2.23157.25 [C:\Program Files\dotnet\sdk]
```
<h4>Summary of project creating:</h4>
<h1 align=center>
 <img align=Center width="30%" src="https://github.com/unoplatform/uno/assets/69314237/e054cf5d-05de-4a56-87a8-9ae75dfe5a4e" />
</h1>

[build output uno platform net 8.txt](https://github.com/unoplatform/uno/files/12002854/build.output.uno.platform.net.8.txt)

**Comment by jeromelaban** [comment]
Thanks for the update. It is likely that we will not provide an update for .NET 8 for the uno.sourcegeneration tasks, since building with  `UnoUIUseRoslynSourceGenerators` set to `true` is a stable feature now.

Why do you need `UnoUIUseRoslynSourceGenerators` to be set to false?

**Reply by limeniye** [reply]
Sorry for my English.
I meant that as default uno-project doesn't has this parameter at .csproj file.

But now I see that it has, but parameter is commented.
What do you think about uncomment this line in order to enable people use Net 8?

**Reply by jeromelaban** [reply]
`UnoUIUseRoslynSourceGenerators` is true by default for new projects and should not be set to false. How are you creating your project?

**Reply by limeniye** [reply]
You was right. **UnoUIUseRoslynSourceGenerators** realy is true as default. But lets see the following test.

Create new Uno Platform Application with following parameters:
- .Net 8;
-  any platforms (In my case I`ve just selected Windows);
- In "Extensions" tab select "Navigation -> Regions".

Try to build your shared project.

Expectation: build should be successed
Result: Xaml generation error

After that try to use the following steps:
- Clean your unsuccessful build;
- Add project parameter ```<UnoUIUseRoslynSourceGenerators>true</UnoUIUseRoslynSourceGenerators>```;
- Rebuild your project.

Result: rebuild will be successfull.

https://github.com/unoplatform/uno/assets/69314237/34434e11-b223-4c60-b4e7-206c58630847

**Reply by jeromelaban** [reply]
Thanks. Indeed, it seems that the .NET 8 SDK is not defining some key property anymore, causing this issue. Let's track in https://github.com/unoplatform/uno/issues/12862
